### PR TITLE
Use generic CollectingNodeVisitor in hql parser

### PR DESCRIPTION
**Pull request by bahusoid**

### CODE REVIEW

File: src/NHibernate/Hql/Ast/ANTLR/QueryTranslatorImpl.cs [hunk 1]
```diff
@@ -89,11 +89,12 @@ public IList List(ISessionImplementor session, QueryParameters queryParameters)
 			ErrorIfDML();
 			var query = ( QueryNode ) _sqlAst;
 			bool hasLimit = queryParameters.RowSelection != null && queryParameters.RowSelection.DefinesLimits;
-			bool needsDistincting = ( query.GetSelectClause().IsDistinct || hasLimit ) && ContainsCollectionFetches;
+			var containsCollectionFetches = ContainsCollectionFetches;
+			bool needsDistincting = ( query.GetSelectClause().IsDistinct || hasLimit ) && containsCollectionFetches;
```
[the remark this answers is not in the document]
Comment (bahusoid) [reply src/NHibernate/Hql/Ast/ANTLR/QueryTranslatorImpl.cs]: yeah.. Better to undo

File: src/NHibernate/Hql/Ast/ANTLR/Tree/FromClause.cs [hunk 2]
```diff
@@ -72,15 +72,27 @@ public FromClause ParentFromClause
...
+		
+		//6.0 TODO: Replace with Typed version below
 		public IList<IASTNode> GetExplicitFromElements()
 		{
-			return ASTUtil.CollectChildren(this, ExplicitFromPredicate);
+			return ASTUtil.CollectChildren<IASTNode>(this, ExplicitFromPredicate);
 		}
-
+		
+		internal IList<FromElement> GetExplicitFromElementsTyped()
```
Comment (hazzik) [inline src/NHibernate/Hql/Ast/ANTLR/Tree/FromClause.cs]: It feels like the return type shall be `IReadOnlyList<>`.

Comment (bahusoid) [reply src/NHibernate/Hql/Ast/ANTLR/Tree/FromClause.cs]: Why? That's not something we care about. Maybe user want to shuffle or slightly modify results in some way before further processing. I see no point in such restriction.

Comment (fredericDelaporte) [reply src/NHibernate/Hql/Ast/ANTLR/Tree/FromClause.cs]: I do not think we should restrict a method return type to be readonly when altering the returned value does not have any impact on the state of the object holding the method.

**Comment by hazzik** [comment]
Ok

>

**Pull request merged**